frontend/src/components/widgets/CameraInput/CameraInput.tsx

```diff
@@ -0,0 +1,579 @@
+/**
+ * @license
+ * Copyright 2018-2021 Streamlit Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from "react"
+import axios from "axios"
+import _ from "lodash"
+
+import { X } from "@emotion-icons/open-iconic"
+import Icon from "src/components/shared/Icon"
+import {
+  FileUploaderState as FileUploaderStateProto,
+  UploadedFileInfo as UploadedFileInfoProto,
+  CameraInput as CameraInputProto,
+} from "src/autogen/proto"
+
+import { FormClearHelper } from "src/components/widgets/Form"
+import { FileUploadClient } from "src/lib/FileUploadClient"
+import { WidgetStateManager } from "src/lib/WidgetStateManager"
+import { logError } from "src/lib/log"
+import {
+  WidgetLabel,
+  StyledWidgetLabelHelp,
+} from "src/components/widgets/BaseWidget"
+import TooltipIcon from "src/components/shared/TooltipIcon"
+import { Placement } from "src/components/shared/Tooltip"
+
+import {
+  UploadFileInfo,
+  UploadedStatus,
+  UploadingStatus,
+} from "../FileUploader/UploadFileInfo"
+
+import CameraInputButton from "./CameraInputButton"
+import WebcamComponent from "./WebcamComponent"
+
+import { StyledBox, StyledCameraInput, StyledSpan } from "./styled-components"
+
+export interface Props {
+  element: CameraInputProto
+  widgetMgr: WidgetStateManager
+  uploadClient: FileUploadClient
+  disabled: boolean
+  width: number
+}
+
+type FileUploaderStatus =
+  | "ready" // FileUploader can upload or delete files
+  | "updating" // at least one file is being uploaded or deleted
+
+interface State {
+  /**
+   * Base64-encoded image data of the current frame from the camera.
+   */
+  imgSrc: string | null
+
+  shutter: boolean
+
+  minShutterEffectPassed: boolean
+  /**
+   * List of files dropped on the FileUploader by the user. This list includes
+   * rejected files that will not be updated.
+   */
+  files: UploadFileInfo[]
```

Comment:
nit: should also rewrite this comment to be less file-uploader specific

Reply:
not a nit: It seems like a lot of the logic below is complicated by keeping this an array to match the `file_uploader` widget. 

Since the camera widget holds at most one uploaded file, I think it'd be worth considering reworking this to no longer use an array here.

Given that this rework has the potential to be a lot of work, I'd be okay with merging this as-is for now to get the widget released, then following up with a refactor quickly.